working/0546-patterns/records-feature-specification.md

```diff
@@ -283,34 +323,88 @@ fields are) and collection literals.
 
 **TODO: Specify this more precisely.**
 
+### Constants
+
+A record expression in a constant context or beginning with `const` defines a
+constant record. A record expression starting with `const` establishes a const
+context for its fields. It is a compile-time error if a field of a constant
+record is not constant.
+
+Since identity is definely loosely for records, an implementation is not
+required to canonicalize equivalent constant records.
+
+```dart
+print(identical(const (1, 2), const (1, 2)));
+```
+
+This may print `true` or `false`.
+
 ## Runtime semantics
 
 ### Records
 
-#### Members
+#### Field getters
 
 Each field in the record's shape exposes a corresponding getter. Invoking that
 getter returns the value provided for that field when the record was created.
 Record fields are immutable and do not have setters.
 
-The `toString()` method's behavior is unspecified.
+#### `toString()`
+
+In debug builds, the `toString()` method converts each field to a string by
+calling `toString()` on its value and prepending it with the field name followed
+by `: ` if the field is named. It concatenates these with `, ` as a separator
+and returns the resulted surrounded by parentheses. For example:
+
+```dart
+print((1, 2, 3).toString()); // "(1, 2, 3)".
+print((a: 'str', 'ing').toString()); // "(a: str, int)".
+```
+
+The order that named fields appear and how they are interleaved with positional
+fields is unspecified. Positional fields must appear in position order. *This
+gives implementations freedom to choose a canonical order for named fields
+independent of the order that the record was created with.*
+
+In a release or optimized build, the behavior of `toString()` is unspecified.
+*This gives implementations freedom to discard the full names of named fields in
+order to reduce code size.* Users should only use `toString()` on records for
+debugging purposes. They are strongly discouraged from parsing the results of
+calling `toString()` or relying on it for end-user visible output.
 
 #### Equality
 
-Records behave similar to other primitive types in Dart with regards to
-equality. They implement `==` such that two records are equal iff they have the
-same shape and all corresponding pairs of fields are equal. Fields are compared
-for equality by calling `==` on the corresponding field values in the same
-order that `==` was called on the records.
+The `==` method on record `r` with right operand `o` is defined as:
+
+1.  If `o` is not a record with the same shape as `r` then return `false`.
+
+1.  For each pair of corresponding positional fields `rf` and `of` in position
+    order:
+
+    1.  If `rf == of` is `false` then return `false`.
```

Comment:
I would like a degree of freedom here that is not allowed.

Consider a record created where the static type is `(Uri, int)`.
This would allow there to be a specialized Record implementation that has a specialized `==` method that knows that `$0` and `$1` have types `Uri` and `int` respectively.

The specification requires `==` to be something like:

```
operator ==(other) =>
    other is (Object?, Object?) &&  // shape test
    this.$0 == other.$0 &&          // Uri.==
    this.$1 == other.$1;            // int.==
```

It would be beneficial to move the faster `int.==` forwards:

```
operator ==(other) =>
    other is (Object?, Object?) &&  // shape test
    this.$1 == other.$1 &&          // int.==
    this.$0 == other.$0;            // Uri.==
```

We can't tell that `Uri.==` is side-effect free - strictly speaking, it does since certain substructures are lazily initialized.
But we do know for sure `int.==` is side-effect free and super efficient.

Can the language be loosened to permit some comparisons to be moved forward provided they have no effects?

Reply:
Simpler statement of my concern: this definition of record equality does not permit `r == r` to take a short-cut.

Reply:
I'd be fine with allowing any order of comparisons. We're going to call *all* of the before answering `true` anyway.

We don't need to promise that we use the same ordering between calls, even on the same shape, because it doesn't matter for the result. And if your `==` has a side effect, ... you had it coming!

Reply:
I'm OK with leaving all field order unspecified. Done.

Reply:
> I'd be fine with allowing any order of comparisons. We're going to call all of the `==`s before answering `true` anyway.

We often accelerate `==` in our own classes via `identical`:

    class Foo {
      bool operator==(other) => identical(this, other) || ... hard way ...;

This is not permitted for records since a field may have a side-effect-free but adversarial `==`.
It seems a shame to do structural equality for the very same allocated record, e.g. looking up an existing record in a Map.

Reply:
Nit-pickily, it's also not permitted/correct for non-record classes that contain `double`s that might be NaN.
We just choose to live with that.
(Records still should't.)